rfc-096-publishing-api-scheduled-publications.md

```diff
@@ -0,0 +1,45 @@
+# Scheduled publication functionality in Publishing API
+
+## Summary
+
+Add scheduled publication functionality to the Publishing API.
+
+## Problem
+
+We have certain scheduled publication formats that we must represent live on GOV.UK within a minute of the scheduled publication time (such as statistical publications).
+
+Scheduled publications are not currently a feature supported by the Publishing API. This means each publishing application that wishes to implement scheduled publishing must do so independently. This leads to duplicate functionality being developed in each publishing application that required the functionality. At the same time, although conceptually each application would develop similar functionality, the code surrounding scheduled publications might vary between applications.
+
+This has several drawbacks:
+
+  * Developers must become familiar with multiple implementations that solve the same problem.
+  * Duplicate functionality across multiple applications violates a Don't Repeat Yourself philosophy.
+  * Debugging problems with scheduled publications across multiple micro service layers is difficult.
+  * There are multiple places where content may get "stuck" and debugging requires knowledge of individual publishing apps.
+
+Whitehall is one of the publishing applications that implements scheduled publishing. In Whitehall's case, the code that interacts with the Publishing API is often difficult to reason about and hard to debug.
+
+## Proposal
+
+Create scheduled publishing functionality in the Publishing API.
+
+- Publishing API MUST provide the ability for publishing apps to schedule a publish of draft content.
+- Publishing API MAY provide a `/v2/content/:content_id/schedule` endpoint.
+- Publishing API MAY introduce a 'scheduled at' date field to editions.
+- Publishing API MAY implement callbacks or push on to message queue.
+- Publishing API MAY implement an endpoint to poll for an item's state (e.g. when in Whitehall admin, as you hit an edition, it polls the Publishing API to check whether it's been published).
+- Publishing API MUST implement a queue of scheduled content (probably use Sidekiq `perform_at` ? as in Whitehall).
+- Publishing API MUST provide the ability to cancel a scheduled publish.
+- Publishing API SHOULD store scheduled requests in the actions log (e.g. `Schedule` and `CancelSchedule`).
+- Publishing API SHOULD allow subsequent requests to overwrite existing scheduled publish requests.
+- Publishing API MUST cancel any scheduled publish requests if the draft edition is discarded.
+- Publishing API MUST cancel any scheduled publish requests if the draft edition is published directly.
+- Whitehall MUST be migrated to use the new Publishing API functionality.
```

Comment:
Note that the way in which Whitehall handles change notes for drafts currently blocks this.

I tripped over this recently, so there's a comment in the code now describing this: https://github.com/alphagov/whitehall/blob/516c8154f19af2a68815cb10cae07b7f87847128/lib/whitehall/publishing_api.rb#L16-L25

Reply:
It's probably also blocked by links in Whitehall too if I remember correctly (as Whitehall uses patch links). So if you were to change links on a document on Whitehall it doesn't apply links until the point of publishing when there is an existed published item otherwise those link changes would affect the published item.